### PR TITLE
#1962: Refactoring of Search Behavior: Restoring of Previous Search Query After Coming Back from Card Interaction or Screen Rotation on Search

### PR DESCRIPTION
**Pull request by vp193dt**
Hello, 
In this pull request, I implemented functionality to save and restore the user's search query during screen rotations and enhance the user experience when navigating back from a selected card, the way user will have the query entered before saved and retrived, based on Issue #1962.
Unit tests were added and all four possible tests were run successfully without any error.

After hours of debugging I figured out the only way this Issue is solveble, where we need to implement 2 Strings to handle the changing text as input from user, because of the behaviour after coming back from a card regarding onQueryTextChange. (alternativelly we could use one string and one boolean for this very same behaviour) - changes in states were not effective on this Issue.
This approach is a bit extraordinary but It solves both parts of problem of this Issue.

**The changes are divided into 2 categories, the first are changes that are connected to screen rotation:**

### Saving Instance State:
I overridden the onSaveInstanceState method. When a user clicks on a card from the list, the current search query (currentQuery) is saved to a variable called finalQuery.
This finalQuery is then stored in the outState bundle, allowing it to be retained even when the screen is rotated.

### Restoring Instance State:
The onRestoreInstanceState method is also overridden to retrieve the saved query when the screen is rotated.
If the savedInstanceState is not null, the stored search query is fetched and assigned to finalQuery.
Finally, if mSearchView is available, the restored query is set in the search view, allowing users to see their previous search term after rotation.

**The second part regarding the retrieving the text after coming back from card picking:**

### New Logic for Search History:
Additional logic was implemented to ensure that the user sees the last search query after returning from a picked card.
If the newText input is empty, the app checks if finalQuery is not empty. If so, it sets the query text to finalQuery for the user.
If both finalQuery and currentQuery are empty, the app clears the current search (currentQuery = "") and resets the search view to show all cards.
When the user changes the text in the search view, the input is stored in currentQuery to be used later for restoring search history.

### Handling Navigation from a Picked Card:
Upon returning from a picked card, if finalQuery is not empty, the search view is expanded to show the previous query.
The finalQuery is reset to an empty string to avoid unintended behavior caused by the onQueryTextChange method, which is called automatically after returning to the search view.

**Tests are focused on both rotation of screen by user (testing same test case as in previous unit test but with rotation commands) and also on simulating steps of the core of the issue with coming back from a card.**

Thank you for feedback.
Have a nice day :)

### CODE REVIEW

**Comment by TheLastProject** [comment]
Thanks for your contribution!

From quick testing I can confirm it does seem to work. However, while reading through your code, I can't help but feel it is very complex for how simple what we're trying to do should be.

While I think the `onSaveInstanceState` and `onRestoreInstanceState` are the correct functions to go to deal with the state loss on rotation, I do quite worry how we went from **one** variable to store the current search filter in (`mFilter`) to **three** variables. I also have my doubts your `onRestoreInstanceState` code does more than set the variable because, from my testing, `mSearchView` seems to always be `null` when in `onRestoreInstanceState`, so the query is never executed.

I've been debugging with the original code (before your change) and I noticed that the value of `mFilter` does correctly stay set even when returning from a card being viewed, so we do actually know the correct search value in `onResume` when we return from the card. However, shortly after, for some reason the `setQuery` gets called on the `mSearchView` and sets `mFilter` to an empty string.

So I guess the real issue here is: why is the `OnQueryTextListener` of the `mSearchView` called with an empty string when we return from the card view activity? I sadly don't really have an answer to this but I think this is the only issue we have to really fix/work around to restore the search when coming back from the card view.

If we have the workaround for restoring the search state from `mFilter` when returning, saving and restoring `mFilter` on rotation (with `onSaveInstanceState/onRestoreInstanceState`) should be all we need to fix both issues.

One thing I think we may want to consider, given [this post I found](https://github.com/material-components/material-components-android/issues/4185#issuecomment-2127448875), is to try to replace the embedded `androidx.appcompat.widget.SearchView` which Catima currently uses with the new Material SearchView in the hope it does not have this `OnQueryTextListener` quirk that's causing all our pain here, which would hopefully allow us to restore it straight from `mFilter`. That would mean leaving this PR alone for a while and focusing on replacing the search view first, which is sadly quite a scope change (created a new issue for that in #2093) :(

---

So, to summarize: I think your change makes sense to fix the issue requested and the general design is in the right direction. They are however complicated due to bugs in the rest of Android. If we replace the SearchView for a more modern component, we may be able to make this PR a lot simpler and cleaner.

It's a pretty decent first PR :)

**Comment by TheLastProject** [comment]
Thinking about this a bit more, and searching the web and finding https://www.reddit.com/r/android_devs/comments/iwwk3x/restoring_the_query_of_a_searchview_after/:

Seeing your change at https://github.com/CatimaLoyalty/Android/pull/2092/files#diff-941e4f5c2de040ee272e9798812bb09cf81bb8a9cd1d03f09ca02a706f7bdf54R608, wouldn't it suffice to start the value of `mFilter` in a new `mFilterToRestore` variable on saving the instance state or when opening another activity to the value of `mFilter` and adding a `setQuery` to your change like:

```
if (mFilterToRestore.isNotEmpty()) {
            searchMenuItem.expandActionView();
            mSearchView.setQuery(mFilterToRestore, false);
            mFilterToRestore = "";
```

If I understood that post correctly, all the other changes would be unneeded.

Just quick brainstorming though, haven't had the time to try this (so, sorry if you tried it already!) But that could bring back the complexity or your PR a lot if it would work. Still annoying to need a second variable so the search view doesn't wipe it due to that one bug but alas.

**Comment by vp193dt** [comment]
Hello, for me this solution did not work when I tested it but please test it in your enviroment so we can be sure, in my case cards remained filtered but search bar was not expanded with search query.
Thanks :)

File: app/src/test/java/protect/card_locker/MainActivityTest.java [hunk 1]
```diff
@@ -443,4 +449,304 @@ public void testFiltering() {
...
+
+        assertTrue(searchMenuItem.isActionViewExpanded());
+        assertEquals("store", mSearchView.getQuery().toString());
+
+        database.close();
+    }
+
+
+
+
```
Comment (TheLastProject) [inline app/src/test/java/protect/card_locker/MainActivityTest.java:751]:
```suggestion
```

File: app/src/test/java/protect/card_locker/MainActivityTest.java [hunk 2]
```diff
@@ -443,4 +449,304 @@ public void testFiltering() {
...
+        MainActivity mainActivity = (MainActivity) activityController.get();
+        activityController.start();
+        activityController.resume();
+
+
+        SQLiteDatabase database = TestHelpers.getEmptyDb(mainActivity).getWritableDatabase();
+        DBHelper.insertLoyaltyCard(database, "The First Store", "Initial note", null, null, new BigDecimal("0"), null, "cardId", null, CatimaBarcode.fromBarcode(BarcodeFormat.UPC_A), Color.BLACK, 0, null,0);
+        DBHelper.insertLoyaltyCard(database, "The Second Store", "Secondary note", null, null, new BigDecimal("0"), null, "cardId", null, CatimaBarcode.fromBarcode(BarcodeFormat.UPC_A), Color.BLACK, 0, null,0);
+
```
Comment (TheLastProject) [inline app/src/test/java/protect/card_locker/MainActivityTest.java:722]:
```suggestion
```

File: app/src/test/java/protect/card_locker/MainActivityTest.java [hunk 3]
```diff
@@ -443,4 +449,304 @@ public void testFiltering() {
...
+    }
+
+    @Test
+    public void testSearchQueryRestorationAfterNavigatingBack() {
+        ActivityController activityController = Robolectric.buildActivity(MainActivity.class).create();
+
+        MainActivity mainActivity = (MainActivity) activityController.get();
+        activityController.start();
+        activityController.resume();
+
```
Comment (TheLastProject) [inline app/src/test/java/protect/card_locker/MainActivityTest.java:717]:
```suggestion
```

File: app/src/test/java/protect/card_locker/MainActivityTest.java [hunk 4]
```diff
@@ -25,6 +30,7 @@
 import org.robolectric.Robolectric;
 import org.robolectric.RobolectricTestRunner;
 import org.robolectric.android.controller.ActivityController;
+import org.robolectric.fakes.RoboMenu;
```
Comment (TheLastProject) [inline app/src/test/java/protect/card_locker/MainActivityTest.java:33]: This import seems unused.

```suggestion
```

File: app/src/test/java/protect/card_locker/MainActivityTest.java [hunk 5]
```diff
@@ -3,18 +3,23 @@
...
 import static org.robolectric.Shadows.shadowOf;
 
 import android.app.Activity;
 import android.content.ComponentName;
 import android.content.SharedPreferences;
+import android.content.res.Configuration;
 import android.database.sqlite.SQLiteDatabase;
 import android.graphics.Color;
 import android.view.Menu;
+import android.view.MenuInflater;
```
Comment (TheLastProject) [inline app/src/test/java/protect/card_locker/MainActivityTest.java:16]: This import seems unused.

```suggestion
```

File: app/src/main/java/protect/card_locker/MainActivity.java [hunk 6]
```diff
@@ -521,6 +542,7 @@ public boolean onCreateOptionsMenu(Menu inputMenu) {
             mSearchView.setSearchableInfo(searchManager.getSearchableInfo(getComponentName()));
             mSearchView.setSubmitButtonEnabled(false);
 
+
```
Comment (TheLastProject) [inline app/src/main/java/protect/card_locker/MainActivity.java:545]:
```suggestion
```

**Review by TheLastProject — COMMENTED** [review]
I tried and tried and can confirm, it really doesn't seem like there is a simpler solution sadly.

I'm still gonna leave this open a bit longer, just in case I get a random eureka moment of something I didn't think of yet, but most likely this will end up being merged, thank you :)

Just a few small comments (and if possible, please revert the whitespace changes, it increases the changes of merge conflicts if the future ever brings a cleaner way, perhaps if we start using another search implementation) :)

**Comment by vp193dt** [comment]
Hello, 
I corrected every suggestion your proposed in my most recent commit so It will be available for you when you would like to merge it to your main.
Thank you.
In case I figures other way of solving this Issue I will let you know immediatelly :)

**Comment by TheLastProject** [comment]
Oh, I actually just found a pretty serious bug in this MR, happens on my Android 14 device :(

1. Search for something
2. Select a card with the search filter active
3. Go back to the main screen
4. Press the little "back" icon in the top-left next to the search -> Search doesn't get dismissed (but keyboard does so that's fine)
5. Press it again -> Search closes but filter stays

https://github.com/user-attachments/assets/eea983fe-6c83-489a-8153-28014f5d9b5c

**Comment by vp193dt** [comment]
Hello,
Yes very concerning bug, that was overlooked unfortunately.
I added clearing of the mFilter in the newest commit and since then could not replicate the bug anymore.
Thank you for warning.

**Comment by TheLastProject** [comment]
I tested almost every combination I could think of and it worked fairly solidly, aside from one last bug I found:
1. Open search
2. Type a search query
3. Rotate screen
4. Click X <- There is nothing in the search field, but the filter is still applied
5. Rotate screen <- Your previous search is back

(So many edge cases, Android's SearchView really is a mess... x.x)

**Comment by vp193dt** [comment]
Hello,
I fixed the bug with the newest commit, I hope it will be now ready for merge, if you find something else let me know, I tried every possible combination and thankfully didnt find any other bug :).
Thank you.

**Comment by TheLastProject** [comment]
Hmm, weird, I tried your latest commit but the bug I described still happens on my side :(

**Comment by vp193dt** [comment]
Hi, 
could you please provide a screen recording of bug? There can be some type of edgecase

**Comment by TheLastProject** [comment]
This is on commit 2847de3e19947ef06fb16ff19d77ac71140a02ce:

https://github.com/user-attachments/assets/d4163942-4122-45bc-8be9-190edae7c36e

**Comment by vp193dt** [comment]
Hello,
so it turned out that I solved another bug I found with my previous commit (And I though it was same as you mentioned),
this was the bug:

1. Open search
2. Type a search query
3. Rotate screen
4. Click 2x back button
5. Rotate screen <- Previous search is back

Now both, also bug you mentioned and recorded the screen for, are fixed, feel free to test it out :)
Thank you.

**Comment by TheLastProject** [comment]
Okay, I'm having serious trouble getting it to misbehave now, so unless I missed something still the code may be good now (even though I'm still worried at how complex this is, I'm hoping that if #2093 happens we can have a much simpler implementation) :smile: 

At this point I'm also not quite sure how useful the tests are, given how much was missed. Is there a reason the tests manually set `mFilter`? Wouldn't it make more sense to use `setQuery` in both tests, like was done in `testSearchQueryRestorationAfterNavigatingBack`? Perhaps we can make the tests more "true to reality" by making that switch. I'm hoping that's the last thing we need to do to make sure this will never break :)

**Comment by vp193dt** [comment]
Hello, 
I edited second test case to avoid using mFilter but in first one, the setting of mFilter is necessary due to unfunctional test when using setQuery in some cases.

**Comment by TheLastProject** [comment]
Sorry for taking so long to respond to this several times, life has been quite hectic. And thanks for all your patience in this *very* long review, this has been quite a challenge to get right so I appreciate you sticking along :)

I've been spending a fair amount of hours seeing if I could get the tests to use the search bar and I couldn't either. I think, however, with how thoroughly we tested this, we should probably just go ahead and merge this because we hopefully caught every possible edge case by now.

I just have one last little thing: I noticed that your new `testSearchHistoryRestorationAfterOrientationChange` test is just a copy-paste of `testFiltering` with a few rotations added. I think we should just add the rotations to `testFiltering` and put a comment there like this:

```
    @Test
    public void testFiltering() {
        // FIXME: This test directly sets mFilter instead of using mSearchView, making it not test the search flow correctly
        // It may falsely succeed or fail, but we're leaving it here so we at least test something instead of nothing
        ActivityController activityController = Robolectric.buildActivity(MainActivity.class).create();

        MainActivity mainActivity = (MainActivity) activityController.get();

        activityController.start();
        activityController.resume();

        LinearLayout helpSection = mainActivity.findViewById(R.id.helpSection);
        TextView noMatchingCardsText = mainActivity.findViewById(R.id.noMatchingCardsText);
        RecyclerView list = mainActivity.findViewById(R.id.list);
        TabLayout groupTabs = mainActivity.findViewById(R.id.groups);

        SQLiteDatabase database = TestHelpers.getEmptyDb(mainActivity).getWritableDatabase();
        DBHelper.insertLoyaltyCard(database, "The First Store", "Initial note", null, null, new BigDecimal("0"), null, "cardId", null, CatimaBarcode.fromBarcode(BarcodeFormat.UPC_A), Color.BLACK, 0, null,0);
        DBHelper.insertLoyaltyCard(database, "The Second Store", "Secondary note", null, null, new BigDecimal("0"), null, "cardId", null, CatimaBarcode.fromBarcode(BarcodeFormat.UPC_A), Color.BLACK, 0, null,0);

        DBHelper.insertGroup(database, "Group one");
        List<Group> groups = new ArrayList<>();
        groups.add(DBHelper.getGroup(database, "Group one"));
        DBHelper.setLoyaltyCardGroups(database, 1, groups);

        activityController.pause();
        activityController.resume();

        assertEquals(View.GONE, helpSection.getVisibility());
        assertEquals(View.GONE, noMatchingCardsText.getVisibility());
        assertEquals(View.VISIBLE, list.getVisibility());

        assertEquals(2, list.getAdapter().getItemCount());

        mainActivity.mFilter = "store";

        activityController.pause();
        activityController.resume();

        Configuration configuration = mainActivity.getResources().getConfiguration();
        configuration.orientation = Configuration.ORIENTATION_LANDSCAPE;
        mainActivity.onConfigurationChanged(configuration);

        configuration.orientation = Configuration.ORIENTATION_PORTRAIT;
        mainActivity.onConfigurationChanged(configuration);

        configuration.orientation = Configuration.ORIENTATION_LANDSCAPE;
        mainActivity.onConfigurationChanged(configuration);

        assertEquals(View.GONE, helpSection.getVisibility());
        assertEquals(View.GONE, noMatchingCardsText.getVisibility());
        assertEquals(View.VISIBLE, list.getVisibility());

        assertEquals(2, list.getAdapter().getItemCount());

        // Switch to Group one
        groupTabs.selectTab(groupTabs.getTabAt(1));

        activityController.pause();
        activityController.resume();

        assertEquals(View.GONE, helpSection.getVisibility());
        assertEquals(View.GONE, noMatchingCardsText.getVisibility());
        assertEquals(View.VISIBLE, list.getVisibility());

        assertEquals(1, list.getAdapter().getItemCount());

        // Switch back to all groups
        groupTabs.selectTab(groupTabs.getTabAt(0));

        assertEquals(View.GONE, helpSection.getVisibility());
        assertEquals(View.GONE, noMatchingCardsText.getVisibility());
        assertEquals(View.VISIBLE, list.getVisibility());

        assertEquals(2, list.getAdapter().getItemCount());

        mainActivity.mFilter = "first";

        activityController.pause();
        activityController.resume();

        assertEquals(View.GONE, helpSection.getVisibility());
        assertEquals(View.GONE, noMatchingCardsText.getVisibility());
        assertEquals(View.VISIBLE, list.getVisibility());

        assertEquals(1, list.getAdapter().getItemCount());

        // Switch to Group one
        groupTabs.selectTab(groupTabs.getTabAt(1));

        activityController.pause();
        activityController.resume();

        assertEquals(View.GONE, helpSection.getVisibility());
        assertEquals(View.GONE, noMatchingCardsText.getVisibility());
        assertEquals(View.VISIBLE, list.getVisibility());

        assertEquals(1, list.getAdapter().getItemCount());

        // Switch back to all groups
        groupTabs.selectTab(groupTabs.getTabAt(0));

        assertEquals(View.GONE, helpSection.getVisibility());
        assertEquals(View.GONE, noMatchingCardsText.getVisibility());
        assertEquals(View.VISIBLE, list.getVisibility());

        assertEquals(1, list.getAdapter().getItemCount());

        mainActivity.mFilter = "initial";

        activityController.pause();
        activityController.resume();

        assertEquals(View.GONE, helpSection.getVisibility());
        assertEquals(View.GONE, noMatchingCardsText.getVisibility());
        assertEquals(View.VISIBLE, list.getVisibility());

        assertEquals(1, list.getAdapter().getItemCount());

        // Switch to Group one
        groupTabs.selectTab(groupTabs.getTabAt(1));

        activityController.pause();
        activityController.resume();

        assertEquals(View.GONE, helpSection.getVisibility());
        assertEquals(View.GONE, noMatchingCardsText.getVisibility());
        assertEquals(View.VISIBLE, list.getVisibility());

        assertEquals(1, list.getAdapter().getItemCount());

        // Switch back to all groups
        groupTabs.selectTab(groupTabs.getTabAt(0));

        assertEquals(View.GONE, helpSection.getVisibility());
        assertEquals(View.GONE, noMatchingCardsText.getVisibility());
        assertEquals(View.VISIBLE, list.getVisibility());

        assertEquals(1, list.getAdapter().getItemCount());

        mainActivity.mFilter = "second";

        configuration.orientation = Configuration.ORIENTATION_LANDSCAPE;
        mainActivity.onConfigurationChanged(configuration);

        configuration.orientation = Configuration.ORIENTATION_PORTRAIT;
        mainActivity.onConfigurationChanged(configuration);

        activityController.pause();
        activityController.resume();

        assertEquals(View.GONE, helpSection.getVisibility());
        assertEquals(View.GONE, noMatchingCardsText.getVisibility());
        assertEquals(View.VISIBLE, list.getVisibility());

        assertEquals(1, list.getAdapter().getItemCount());

        // Switch to Group one
        groupTabs.selectTab(groupTabs.getTabAt(1));

        activityController.pause();
        activityController.resume();

        shadowOf(getMainLooper()).idle();

        assertEquals(View.GONE, helpSection.getVisibility());
        assertEquals(View.VISIBLE, noMatchingCardsText.getVisibility());
        assertEquals(View.GONE, list.getVisibility());

        assertEquals(0, list.getAdapter().getItemCount());

        // Switch back to all groups
        groupTabs.selectTab(groupTabs.getTabAt(0));

        assertEquals(View.GONE, helpSection.getVisibility());
        assertEquals(View.GONE, noMatchingCardsText.getVisibility());
        assertEquals(View.VISIBLE, list.getVisibility());

        assertEquals(1, list.getAdapter().getItemCount());

        mainActivity.mFilter = "company";

        // Rotate to landscape (right)
        configuration.orientation = Configuration.ORIENTATION_LANDSCAPE;
        mainActivity.onConfigurationChanged(configuration);

        activityController.pause();
        activityController.resume();

        shadowOf(getMainLooper()).idle();

        assertEquals(View.GONE, helpSection.getVisibility());
        assertEquals(View.VISIBLE, noMatchingCardsText.getVisibility());
        assertEquals(View.GONE, list.getVisibility());

        assertEquals(0, list.getAdapter().getItemCount());

        // Switch to Group one
        groupTabs.selectTab(groupTabs.getTabAt(1));

        activityController.pause();
        activityController.resume();

        shadowOf(getMainLooper()).idle();

        assertEquals(View.GONE, helpSection.getVisibility());
        assertEquals(View.VISIBLE, noMatchingCardsText.getVisibility());
        assertEquals(View.GONE, list.getVisibility());

        assertEquals(0, list.getAdapter().getItemCount());

        // Switch back to all groups
        groupTabs.selectTab(groupTabs.getTabAt(0));

        assertEquals(View.GONE, helpSection.getVisibility());
        assertEquals(View.VISIBLE, noMatchingCardsText.getVisibility());
        assertEquals(View.GONE, list.getVisibility());

        assertEquals(0, list.getAdapter().getItemCount());

        mainActivity.mFilter = "";

        activityController.pause();
        activityController.resume();

        shadowOf(getMainLooper()).idle();

        assertEquals(View.GONE, helpSection.getVisibility());
        assertEquals(View.GONE, noMatchingCardsText.getVisibility());
        assertEquals(View.VISIBLE, list.getVisibility());

        assertEquals(2, list.getAdapter().getItemCount());

        // Switch to Group one
        groupTabs.selectTab(groupTabs.getTabAt(1));

        activityController.pause();
        activityController.resume();

        shadowOf(getMainLooper()).idle();

        assertEquals(View.GONE, helpSection.getVisibility());
        assertEquals(View.GONE, noMatchingCardsText.getVisibility());
        assertEquals(View.VISIBLE, list.getVisibility());

        assertEquals(1, list.getAdapter().getItemCount());

        // Switch back to all groups
        groupTabs.selectTab(groupTabs.getTabAt(0));

        assertEquals(View.GONE, helpSection.getVisibility());
        assertEquals(View.GONE, noMatchingCardsText.getVisibility());
        assertEquals(View.VISIBLE, list.getVisibility());

        assertEquals(2, list.getAdapter().getItemCount());

        database.close();
    }
```

**Comment by vp193dt** [comment]
Hello,
nice to hear that it is ready to merge :),
I edited the testcases, everythings ready now I think,
Thank you

**Comment by TheLastProject** [comment]
Thanks! I rebased and squashed it before merging, but it's in now :)
